packages/netlify-cms-core/src/lib/registry.js

```diff
@@ -106,7 +115,19 @@ export function resolveWidget(name) {
  */
 export function registerEditorComponent(component) {
   const plugin = EditorComponent(component);
-  registry.editorComponents = registry.editorComponents.set(plugin.get('id'), plugin);
+  if (plugin.type === 'code-block') {
+    const codeBlock = registry.editorComponents.find(c => c.type === 'code-block');
+
+    if (codeBlock) {
+      console.warn(oneLine`
+        Only one editor component of type "code-block" may be registered. Previously registered code
+        block component(s) will be overwritten.
+      `);
+      registry.editorComponents = registry.editorComponents.delete(codeBlock.id);
+    }
+  }
+
+  registry.editorComponents = registry.editorComponents.set(plugin.id, plugin);
```

Comment:
Introduces an experimental API for editor components that integrate deep with Slate for specific roles, starting with the **code-block** type. This approach allows any editor component to utilize the integration.